File: .agents/skills/attribute-ai-commits/scripts/coauthor-registry.mjs
#!/usr/bin/env node

import { readFileSync } from "node:fs";
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";

const scriptDirectory = dirname(fileURLToPath(import.meta.url));
const defaultRegistry = resolve(scriptDirectory, "../references/providers.json");
const providerPattern = /^[a-z0-9]+(?:-[a-z0-9]+)*$/;
const repositoryPattern = /^[A-Za-z0-9_.-]+\/[A-Za-z0-9_.-]+$/;
const commitPattern = /^[0-9a-fA-F]{40}$/;

class RegistryError extends Error {}

function output(value) {
  process.stdout.write(`${JSON.stringify(value, null, 2)}\n`);
}

function loadRegistry(path) {
  let registry;
  try {
    registry = JSON.parse(readFileSync(path, "utf8"));
  } catch (error) {
    throw new RegistryError(`could not read registry ${path}: ${error.message}`);
  }
  if (
    registry.schema !== 1 ||
    !registry.providers ||
    typeof registry.providers !== "object" ||
    Array.isArray(registry.providers)
  ) {
    throw new RegistryError(`unsupported registry schema in ${path}`);
  }
  if (!registry.scope || !registry.as_of) {
    throw new RegistryError(`registry scope or as_of date missing in ${path}`);
  }
  return registry;
}

function resolveProvider(registry, requested) {
  const normalized = requested.toLocaleLowerCase();
  for (const [provider, entry] of Object.entries(registry.providers)) {
    const names = [provider, ...(entry.aliases ?? [])].map((name) => name.toLocaleLowerCase());
    if (names.includes(normalized)) return [provider, entry];
  }
  return null;
}

function validateRegistry(registry) {
  const seenNames = new Map();
  const counts = { verified: 0, unavailable: 0 };
  for (const [provider, entry] of Object.entries(registry.providers)) {
    if (!providerPattern.test(provider)) throw new RegistryError(`invalid provider slug: ${provider}`);
    if (!Object.hasOwn(counts, entry.status)) {
      throw new RegistryError(`invalid status for ${provider}: ${entry.status}`);
    }
    counts[entry.status] += 1;
    for (const name of [provider, ...(entry.aliases ?? [])]) {
      const normalized = name.toLocaleLowerCase();
      if (seenNames.has(normalized)) {
        throw new RegistryError(`duplicate provider name ${name}: ${provider} and ${seenNames.get(normalized)}`);
      }
      seenNames.set(normalized, provider);
    }
    if (entry.status === "verified") {
      for (const field of ["label", "github_login", "email", "verification"]) {
        if (!entry[field]) throw new RegistryError(`verified provider ${provider} lacks ${field}`);
      }
      const verification = entry.verification;
      if (!["github_profile", "rendered_commit"].includes(verification.kind)) {
        throw new RegistryError(`invalid verification kind for ${provider}`);
      }
      if (verification.kind === "github_profile" && !Number.isInteger(verification.id)) {
        throw new RegistryError(`invalid GitHub profile proof for ${provider}`);
      }
      if (verification.kind === "rendered_commit") {
        const parts = verification.commit_parts;
        const validParts =
          Array.isArray(parts) &&
          parts.length === 2 &&
          parts.every((part) => typeof part === "string" && part.length === 20) &&
          commitPattern.test(parts.join(""));
        if (!repositoryPattern.test(verification.repository ?? "") || !validParts) {
          throw new RegistryError(`invalid rendered commit proof for ${provider}`);
        }
      }
    } else {
      for (const field of ["reason", "evidence", "checked_at"]) {
        if (!entry[field]) throw new RegistryError(`unavailable provider ${provider} lacks ${field}`);
      }
    }
  }
  return counts;
}

function parseArguments(arguments_) {
  const values = [...arguments_];
  let registryPath = defaultRegistry;
  if (values[0] === "--registry") {
    if (!values[1]) throw new RegistryError("--registry requires a path");
    registryPath = resolve(values[1]);
    values.splice(0, 2);
  }
  return { registryPath, command: values[0], operands: values.slice(1) };
}

function main() {
  const { registryPath, command, operands } = parseArguments(process.argv.slice(2));
  const registry = loadRegistry(registryPath);
  const counts = validateRegistry(registry);
  if (command === "list" && operands.length === 0) {
    output(
      Object.entries(registry.providers).map(([agent, entry]) => ({
        agent,
        status: entry.status,
        label: entry.label,
      }))
    );
    return 0;
  }
  if (command === "validate" && operands.length === 0) {
    output({ status: "valid", ...counts, total: counts.verified + counts.unavailable });
    return 0;
  }
  if (command === "lookup" && operands.length === 1) {
    const resolvedProvider = resolveProvider(registry, operands[0]);
    if (!resolvedProvider) {
      output({ agent: operands[0], status: "missing" });
      return 3;
    }
    const [agent, entry] = resolvedProvider;
    const result = { agent, ...entry };
    if (entry.verification?.kind === "rendered_commit") {
      result.proof_url = `https://github.com/${entry.verification.repository}/commit/${entry.verification.commit_parts.join("")}`;
    }
    output(result);
    return entry.status === "verified" ? 0 : 2;
  }
  throw new RegistryError("usage: coauthor-registry.mjs [--registry <path>] list|validate|lookup <agent>");
}

try {
  process.exitCode = main();
} catch (error) {
  output({ status: "error", reason: error instanceof Error ? error.message : String(error) });
  process.exitCode = 1;
}

File: .agents/skills/attribute-ai-commits/scripts/coauthor_registry.py
#!/usr/bin/env python3
"""Read the co-author registry or verify a candidate GitHub identity."""

from __future__ import annotations

import argparse
import html
import json
import re
import sys
import urllib.error
import urllib.parse
import urllib.request
from datetime import date
from pathlib import Path
from typing import Any


REGISTRY = Path(__file__).resolve().parent.parent / "references" / "providers.json"
USER_AGENT = "a-novel-ai-coauthor-verifier/1"
NOREPLY_PATTERN = re.compile(
    r"^(?P<id>[0-9]+)\+(?P<login>.+)@users\.noreply\.github\.com$",
    re.IGNORECASE,
)
GITHUB_COMMIT_PATTERN = re.compile(
    r"^https://github\.com/(?P<repository>[A-Za-z0-9_.-]+/[A-Za-z0-9_.-]+)/commit/"
    r"(?P<commit>[0-9a-fA-F]{40})/?$"
)


class RegistryError(RuntimeError):
    """Report an operational error separately from an unavailable identity."""


def read_json(url: str) -> dict[str, Any]:
    request = urllib.request.Request(
        url,
        headers={"Accept": "application/vnd.github+json", "User-Agent": USER_AGENT},
    )
    try:
        with urllib.request.urlopen(request, timeout=20) as response:
            return json.load(response)
    except (OSError, urllib.error.URLError, json.JSONDecodeError) as error:
        raise RegistryError(f"could not read {url}: {error}") from error


def read_text(url: str) -> str:
    request = urllib.request.Request(url, headers={"User-Agent": USER_AGENT})
    try:
        with urllib.request.urlopen(request, timeout=20) as response:
            return response.read().decode("utf-8")
    except (OSError, urllib.error.URLError, UnicodeDecodeError) as error:
        raise RegistryError(f"could not read {url}: {error}") from error


def load_registry(path: Path) -> dict[str, Any]:
    try:
        registry = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError) as error:
        raise RegistryError(f"could not read registry {path}: {error}") from error
    if registry.get("schema") != 1 or not isinstance(registry.get("providers"), dict):
        raise RegistryError(f"unsupported registry schema in {path}")
    if not registry.get("scope") or not registry.get("as_of"):
        raise RegistryError(f"registry scope or as_of date missing in {path}")
    return registry


def resolve_provider(registry: dict[str, Any], requested: str) -> tuple[str, dict[str, Any]] | None:
    normalized = requested.casefold()
    for provider, entry in registry["providers"].items():
        aliases = [provider, *entry.get("aliases", [])]
        if normalized in {alias.casefold() for alias in aliases}:
            return provider, entry
    return None


def validate_registry(registry: dict[str, Any]) -> dict[str, int]:
    seen_names: dict[str, str] = {}
    counts = {"verified": 0, "unavailable": 0}
    for provider, entry in registry["providers"].items():
        if not re.fullmatch(r"[a-z0-9]+(?:-[a-z0-9]+)*", provider):
            raise RegistryError(f"invalid provider slug: {provider}")
        status = entry.get("status")
        if status not in counts:
            raise RegistryError(f"invalid status for {provider}: {status}")
        counts[status] += 1
        for name in [provider, *entry.get("aliases", [])]:
            normalized = name.casefold()
            if normalized in seen_names:
                raise RegistryError(
                    f"duplicate provider name {name}: {provider} and {seen_names[normalized]}"
                )
            seen_names[normalized] = provider
        if status == "verified":
            for field in ("label", "github_login", "email", "verification"):
                if not entry.get(field):
                    raise RegistryError(f"verified provider {provider} lacks {field}")
            verification = entry["verification"]
            kind = verification.get("kind")
            if kind not in {"github_profile", "rendered_commit"}:
                raise RegistryError(f"invalid verification kind for {provider}")
            if kind == "github_profile" and not isinstance(verification.get("id"), int):
                raise RegistryError(f"invalid GitHub profile proof for {provider}")
            if kind == "rendered_commit":
                repository = verification.get("repository", "")
                parts = verification.get("commit_parts")
                if not re.fullmatch(
                    r"[A-Za-z0-9_.-]+/[A-Za-z0-9_.-]+",
                    repository,
                ) or not (
                    isinstance(parts, list)
                    and len(parts) == 2
                    and all(isinstance(part, str) and len(part) == 20 for part in parts)
                    and re.fullmatch(r"[0-9a-fA-F]{40}", "".join(parts))
                ):
                    raise RegistryError(f"invalid rendered commit proof for {provider}")
        else:
            for field in ("reason", "evidence", "checked_at"):
                if not entry.get(field):
                    raise RegistryError(f"unavailable provider {provider} lacks {field}")
    return counts


def rendered_authors(document: str) -> list[dict[str, Any]]:
    for match in re.finditer(r'"authors":(\[[^]]*\])', document):
        try:
            authors = json.loads(match.group(1))
        except json.JSONDecodeError:
            continue
        if isinstance(authors, list):
            return [author for author in authors if isinstance(author, dict)]
    raise RegistryError("GitHub page did not expose a rendered authors record")


def verify_profile(login: str, email: str) -> tuple[bool, dict[str, Any]]:
    match = NOREPLY_PATTERN.fullmatch(email)
    if not match:
        raise RegistryError("provider-domain emails require --proof-commit")
    quoted_login = urllib.parse.quote(login, safe="")
    profile = read_json(f"https://api.github.com/users/{quoted_login}")
    expected_id = int(match.group("id"))
    valid = (
        profile.get("login", "").casefold() == login.casefold()
        and profile.get("id") == expected_id
        and match.group("login").casefold() == login.casefold()
    )
    return valid, {"kind": "github_profile", "id": expected_id}


def compact_proof_commit(url: str) -> dict[str, Any]:
    match = GITHUB_COMMIT_PATTERN.fullmatch(url)
    if not match:
        raise RegistryError("proof commit must be a canonical GitHub commit URL with a full hash")
    commit = match.group("commit").lower()
    return {
        "kind": "rendered_commit",
        "repository": match.group("repository"),
        "commit_parts": [commit[:20], commit[20:]],
    }


def verify_proof_commit(login: str, email: str, url: str) -> tuple[bool, dict[str, Any]]:
    verification = compact_proof_commit(url)
    document = read_text(url)
    readable_document = html.unescape(document)
    trailer = re.compile(
        rf"Co-authored-by:[^\n<]*<{re.escape(email)}>",
        re.IGNORECASE,
    )
    if not trailer.search(readable_document):
        raise RegistryError("proof commit does not contain the candidate email")
    expected_login = login.casefold()
    valid = any(
        str(author.get("login", "")).casefold() == expected_login
        for author in rendered_authors(document)
    )
    return valid, verification


def command_list(args: argparse.Namespace) -> int:
    registry = load_registry(args.registry)
    validate_registry(registry)
    listing = [
        {"agent": provider, "status": entry["status"], "label": entry.get("label")}
        for provider, entry in registry["providers"].items()
    ]
    print(json.dumps(listing, indent=2, sort_keys=True))
    return 0


def command_validate(args: argparse.Namespace) -> int:
    registry = load_registry(args.registry)
    counts = validate_registry(registry)
    output = {"status": "valid", **counts, "total": sum(counts.values())}
    print(json.dumps(output, indent=2, sort_keys=True))
    return 0


def command_lookup(args: argparse.Namespace) -> int:
    registry = load_registry(args.registry)
    resolved = resolve_provider(registry, args.agent)
    if resolved is None:
        print(json.dumps({"agent": args.agent, "status": "missing"}, indent=2, sort_keys=True))
        return 3
    provider, entry = resolved
    output = {"agent": provider, **entry}
    verification = output.get("verification", {})
    if verification.get("kind") == "rendered_commit":
        output["proof_url"] = (
            f"https://github.com/{verification['repository']}/commit/"
            f"{''.join(verification['commit_parts'])}"
        )
    print(json.dumps(output, indent=2, sort_keys=True))
    return 0 if entry.get("status") == "verified" else 2


def command_verify(args: argparse.Namespace) -> int:
    if NOREPLY_PATTERN.fullmatch(args.email) and not args.proof_commit:
        valid, verification = verify_profile(args.github_login, args.email)
    elif args.proof_commit:
        valid, verification = verify_proof_commit(
            args.github_login,
            args.email,
            args.proof_commit,
        )
    else:
        raise RegistryError("cannot prove this email without --proof-commit")

    verification["verified_at"] = date.today().isoformat()
    entry = {
        args.agent: {
            "aliases": [],
            "status": "verified" if valid else "unavailable",
            "label": args.label,
            "github_login": args.github_login,
            "email": args.email,
            "verification": verification,
        }
    }
    if not valid:
        entry[args.agent]["reason"] = "GitHub did not resolve the expected account"
    print(json.dumps(entry, indent=2, sort_keys=True))
    return 0 if valid else 2


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--registry", type=Path, default=REGISTRY, help=argparse.SUPPRESS)
    commands = parser.add_subparsers(dest="command", required=True)

    list_command = commands.add_parser("list", help="list every registered agent")
    list_command.set_defaults(handler=command_list)

    validate = commands.add_parser(
        "validate",
        help="validate the full registry without network access",
    )
    validate.set_defaults(handler=command_validate)

    lookup = commands.add_parser("lookup", help="read a registered agent without network access")
    lookup.add_argument("agent")
    lookup.set_defaults(handler=command_lookup)

    verify = commands.add_parser("verify", help="verify a missing agent candidate")
    verify.add_argument("--agent", required=True)
    verify.add_argument("--label", required=True)
    verify.add_argument("--github-login", required=True)
    verify.add_argument("--email", required=True)
    verify.add_argument("--proof-commit")
    verify.set_defaults(handler=command_verify)
    return parser.parse_args()


def main() -> int:
    args = parse_args()
    try:
        return args.handler(args)
    except RegistryError as error:
        print(
            json.dumps(
                {"status": "error", "reason": str(error)},
                indent=2,
                sort_keys=True,
            )
        )
        return 1


if __name__ == "__main__":
    sys.exit(main())
